editor.tsx

```diff
@@ -3,29 +3,104 @@ import { ASSET_CACHE_BUST_KEY } from "$fresh/runtime.ts";
 import { renderToString } from "preact-render-to-string";
 import LiveContext from "./context.ts";
 import getSupabaseClient, { getSupabaseClientForUser } from "./supabase.ts";
-import { Flag, Module } from "./types.ts";
+import { Flag, Module, PageComponentData, PageLoaderData } from "./types.ts";
 import {
   componentNameFromPath,
   getComponentModule,
 } from "./utils/component.ts";
 import { createServerTiming } from "./utils/serverTimings.ts";
 import { duplicateProdPage, getFlagFromPageId } from "./utils/supabase.ts";
+import { JSONSchema7 } from "https://esm.sh/v92/@types/json-schema@7.0.11/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0CmQvcHJlYWN0QDEwLjEwLjY/index.d.ts";
+import {
+  generateLoaderInstance,
+  loaderInstanceToProp,
+  loaderPathToKey,
+} from "./utils/loaders.ts";
+import { LivePageData } from "./live.tsx";
 
 const ONE_YEAR_CACHE = "public, max-age=31536000, immutable";
 
-const updateDraft = async (req: Request, url: URL, ctx: any) => {
-  const { components, template, siteId, variantId, experiment } = ctx;
+/**
+ * @description This function creates a loaders list based on component props that depends on Loaders.
+ *  Also, mutate prop that depends on Loaders to component[index].props.[propName] = `loadername-uuid`
+ */
+const generateLoadersFromComponents = (
+  _: Request,
+  __: URL,
+  ctx: any,
+): PageLoaderData[] => {
+  const { components } = ctx;
+  const loaders: PageLoaderData[] = [];
+  const manifest = LiveContext.getManifest();
+
+  for (const component of components) {
+    const componentSchema = manifest.schemas[component.component];
+
+    if (!componentSchema) {
+      continue;
+    }
+
+    const propsThatRequireLoader = Object.entries(
+      componentSchema.properties ?? {},
+    )
+      .filter(([, value]) => Boolean((value as JSONSchema7).$ref));
+
+    if (propsThatRequireLoader.length === 0) {
+      continue;
+    }
+
+    propsThatRequireLoader.forEach(([property, propertySchema]) => {
+      if (typeof propertySchema !== "object") {
+        return;
+      }
+
+      if (!propertySchema?.$ref) {
+        throw new Error(`Property ${property} doesn't have $ref on schema`);
+      }
+
+      // Match property ref input into loader output
+      const [loaderPath] = Object.entries(manifest.loaders).find((
+        [, loader],
+      ) => loader.default.outputSchema.$ref === propertySchema.$ref) ?? [];
+
+      if (!loaderPath) {
+        throw new Error(
+          `Doesn't exists loader with this $ref: ${propertySchema.$ref}`,
+        );
+      }
+
+      const loaderProp = component.props[property];
+      const loaderInstanceName = generateLoaderInstance(propertySchema.$ref);
+      component.props[property] = loaderInstanceToProp(loaderInstanceName);
```

Comment:
Can we do it in a way that we return a new array of components instead of mutation the object? It's usually a bit safer.

I'm not sure, depends on how we're using this data.

Reply:
Yeah, I totally agree with it. I was thinking using the `structuredClone` to clone the whole components array, then mutate it and return { components, loaders }